File: MuscleHamster/Models/AudioPreferences.swift
//
//  AudioPreferences.swift
//  MuscleHamster
//
//  Audio settings model with UserDefaults persistence
//  Phase 08.1: Audio Experience and Settings
//

import Foundation
import SwiftUI

// MARK: - Audio Preferences Keys

/// UserDefaults keys for audio settings persistence
enum AudioPreferencesKey {
    static let globalMute = "audio_globalMute"
    static let sfxEnabled = "audio_sfxEnabled"
    static let musicEnabled = "audio_musicEnabled"
    static let sfxVolume = "audio_sfxVolume"
    static let musicVolume = "audio_musicVolume"
    static let mixWithOthers = "audio_mixWithOthers"
}

// MARK: - Audio Preferences

/// User's audio preferences with persistence
/// Access via @AppStorage in views or AudioManager.shared.preferences
struct AudioPreferences {

    // MARK: - Default Values

    static let defaultGlobalMute = false
    static let defaultSfxEnabled = true
    static let defaultMusicEnabled = true
    static let defaultSfxVolume: Float = 0.8
    static let defaultMusicVolume: Float = 0.5
    static let defaultMixWithOthers = true

    // MARK: - Properties

    /// Master mute switch - disables all app audio
    var globalMute: Bool

    /// Sound effects enabled (workout cues, celebrations, UI sounds)
    var soundEffectsEnabled: Bool

    /// Background music enabled
    var musicEnabled: Bool

    /// Sound effects volume (0.0 - 1.0)
    var soundEffectsVolume: Float

    /// Music volume (0.0 - 1.0)
    var musicVolume: Float

    /// Mix with other apps (allows podcasts/music to play alongside)
    var mixWithOthers: Bool

    // MARK: - Computed Properties

    /// Whether sound effects can currently play
    var canPlaySFX: Bool {
        !globalMute && soundEffectsEnabled
    }

    /// Whether music can currently play
    var canPlayMusic: Bool {
        !globalMute && musicEnabled
    }

    /// Effective SFX volume (0 if disabled)
    var effectiveSFXVolume: Float {
        canPlaySFX ? soundEffectsVolume : 0
    }

    /// Effective music volume (0 if disabled)
    var effectiveMusicVolume: Float {
        canPlayMusic ? musicVolume : 0
    }

    /// SFX volume as percentage (0-100)
    var sfxVolumePercent: Int {
        Int(soundEffectsVolume * 100)
    }

    /// Music volume as percentage (0-100)
    var musicVolumePercent: Int {
        Int(musicVolume * 100)
    }

    // MARK: - Initialization

    /// Creates preferences with specified values
    init(
        globalMute: Bool = defaultGlobalMute,
        soundEffectsEnabled: Bool = defaultSfxEnabled,
        musicEnabled: Bool = defaultMusicEnabled,
        soundEffectsVolume: Float = defaultSfxVolume,
        musicVolume: Float = defaultMusicVolume,
        mixWithOthers: Bool = defaultMixWithOthers
    ) {
        self.globalMute = globalMute
        self.soundEffectsEnabled = soundEffectsEnabled
        self.musicEnabled = musicEnabled
        self.soundEffectsVolume = max(0, min(1, soundEffectsVolume))
        self.musicVolume = max(0, min(1, musicVolume))
        self.mixWithOthers = mixWithOthers
    }

    /// Loads preferences from UserDefaults
    static func load() -> AudioPreferences {
        let defaults = UserDefaults.standard

        return AudioPreferences(
            globalMute: defaults.bool(forKey: AudioPreferencesKey.globalMute),
            soundEffectsEnabled: defaults.object(forKey: AudioPreferencesKey.sfxEnabled) as? Bool ?? defaultSfxEnabled,
            musicEnabled: defaults.object(forKey: AudioPreferencesKey.musicEnabled) as? Bool ?? defaultMusicEnabled,
            soundEffectsVolume: defaults.object(forKey: AudioPreferencesKey.sfxVolume) as? Float ?? defaultSfxVolume,
            musicVolume: defaults.object(forKey: AudioPreferencesKey.musicVolume) as? Float ?? defaultMusicVolume,
            mixWithOthers: defaults.object(forKey: AudioPreferencesKey.mixWithOthers) as? Bool ?? defaultMixWithOthers
        )
    }

    /// Saves preferences to UserDefaults
    func save() {
        let defaults = UserDefaults.standard
        defaults.set(globalMute, forKey: AudioPreferencesKey.globalMute)
        defaults.set(soundEffectsEnabled, forKey: AudioPreferencesKey.sfxEnabled)
        defaults.set(musicEnabled, forKey: AudioPreferencesKey.musicEnabled)
        defaults.set(soundEffectsVolume, forKey: AudioPreferencesKey.sfxVolume)
        defaults.set(musicVolume, forKey: AudioPreferencesKey.musicVolume)
        defaults.set(mixWithOthers, forKey: AudioPreferencesKey.mixWithOthers)
    }
}

// MARK: - Sound Effect Types

/// Available sound effect types in the app
enum SoundEffect: String, CaseIterable {
    case buttonTap = "button_tap"
    case workoutStart = "workout_start"
    case workoutComplete = "workout_complete"
    case exerciseTransition = "exercise_transition"
    case celebration = "celebration"
    case pointsEarned = "points_earned"
    case streakIncrement = "streak_increment"
    case hamsterHappy = "hamster_happy"
    case hamsterExcited = "hamster_excited"
    case purchase = "purchase"
    case levelUp = "level_up"

    /// Display name for the sound effect
    var displayName: String {
        switch self {
        case .buttonTap: return "Button Tap"
        case .workoutStart: return "Workout Start"
        case .workoutComplete: return "Workout Complete"
        case .exerciseTransition: return "Exercise Transition"
        case .celebration: return "Celebration"
        case .pointsEarned: return "Points Earned"
        case .streakIncrement: return "Streak Up"
        case .hamsterHappy: return "Happy Hamster"
        case .hamsterExcited: return "Excited Hamster"
        case .purchase: return "Purchase"
        case .levelUp: return "Level Up"
        }
    }

    /// File extension (all sounds are .wav for low latency)
    var fileExtension: String { "wav" }

    /// Bundle resource name
    var resourceName: String { rawValue }
}

// MARK: - Music Track Types

/// Available music tracks in the app
enum MusicTrack: String, CaseIterable {
    case workoutUpbeat = "workout_upbeat"
    case workoutChill = "workout_chill"
    case workoutIntense = "workout_intense"
    case ambient = "ambient"

    /// Display name for the track
    var displayName: String {
        switch self {
        case .workoutUpbeat: return "Upbeat"
        case .workoutChill: return "Chill"
        case .workoutIntense: return "Intense"
        case .ambient: return "Ambient"
        }
    }

    /// Track description
    var description: String {
        switch self {
        case .workoutUpbeat: return "Energetic and motivating"
        case .workoutChill: return "Relaxed and steady"
        case .workoutIntense: return "High energy for tough workouts"
        case .ambient: return "Calm background sounds"
        }
    }

    /// File extension
    var fileExtension: String { "mp3" }

    /// Bundle resource name
    var resourceName: String { rawValue }
}

File: MuscleHamsterExpo/src/models/AudioPreferences.js
// Audio Preferences Model
// Ported from Swift AudioPreferences.swift
// Phase 08.1: Audio Experience and Settings

import AsyncStorage from '@react-native-async-storage/async-storage';
import Logger from '../services/LoggerService';

// Storage Keys
const AUDIO_STORAGE_KEYS = {
  globalMute: '@audio_globalMute',
  sfxEnabled: '@audio_sfxEnabled',
  musicEnabled: '@audio_musicEnabled',
  sfxVolume: '@audio_sfxVolume',
  musicVolume: '@audio_musicVolume',
  mixWithOthers: '@audio_mixWithOthers',
};

// Default Values
export const AUDIO_DEFAULTS = {
  globalMute: false,
  soundEffectsEnabled: true,
  musicEnabled: true,
  soundEffectsVolume: 0.8,
  musicVolume: 0.5,
  mixWithOthers: true,
};

// Sound Effect Types
export const SoundEffect = {
  BUTTON_TAP: 'button_tap',
  WORKOUT_START: 'workout_start',
  WORKOUT_COMPLETE: 'workout_complete',
  EXERCISE_TRANSITION: 'exercise_transition',
  CELEBRATION: 'celebration',
  POINTS_EARNED: 'points_earned',
  STREAK_INCREMENT: 'streak_increment',
  HAMSTER_HAPPY: 'hamster_happy',
  HAMSTER_EXCITED: 'hamster_excited',
  PURCHASE: 'purchase',
  LEVEL_UP: 'level_up',
};

export const SoundEffectInfo = {
  [SoundEffect.BUTTON_TAP]: { displayName: 'Button Tap', fileExtension: 'wav' },
  [SoundEffect.WORKOUT_START]: { displayName: 'Workout Start', fileExtension: 'wav' },
  [SoundEffect.WORKOUT_COMPLETE]: { displayName: 'Workout Complete', fileExtension: 'wav' },
  [SoundEffect.EXERCISE_TRANSITION]: { displayName: 'Exercise Transition', fileExtension: 'wav' },
  [SoundEffect.CELEBRATION]: { displayName: 'Celebration', fileExtension: 'wav' },
  [SoundEffect.POINTS_EARNED]: { displayName: 'Points Earned', fileExtension: 'wav' },
  [SoundEffect.STREAK_INCREMENT]: { displayName: 'Streak Up', fileExtension: 'wav' },
  [SoundEffect.HAMSTER_HAPPY]: { displayName: 'Happy Hamster', fileExtension: 'wav' },
  [SoundEffect.HAMSTER_EXCITED]: { displayName: 'Excited Hamster', fileExtension: 'wav' },
  [SoundEffect.PURCHASE]: { displayName: 'Purchase', fileExtension: 'wav' },
  [SoundEffect.LEVEL_UP]: { displayName: 'Level Up', fileExtension: 'wav' },
};

// Music Track Types
export const MusicTrack = {
  WORKOUT_UPBEAT: 'workout_upbeat',
  WORKOUT_CHILL: 'workout_chill',
  WORKOUT_INTENSE: 'workout_intense',
  AMBIENT: 'ambient',
};

export const MusicTrackInfo = {
  [MusicTrack.WORKOUT_UPBEAT]: {
    displayName: 'Upbeat',
    description: 'Energetic and motivating',
    fileExtension: 'mp3',
  },
  [MusicTrack.WORKOUT_CHILL]: {
    displayName: 'Chill',
    description: 'Relaxed and steady',
    fileExtension: 'mp3',
  },
  [MusicTrack.WORKOUT_INTENSE]: {
    displayName: 'Intense',
    description: 'High energy for tough workouts',
    fileExtension: 'mp3',
  },
  [MusicTrack.AMBIENT]: {
    displayName: 'Ambient',
    description: 'Calm background sounds',
    fileExtension: 'mp3',
  },
};

// Create default audio preferences
export const createDefaultAudioPreferences = () => ({
  globalMute: AUDIO_DEFAULTS.globalMute,
  soundEffectsEnabled: AUDIO_DEFAULTS.soundEffectsEnabled,
  musicEnabled: AUDIO_DEFAULTS.musicEnabled,
  soundEffectsVolume: AUDIO_DEFAULTS.soundEffectsVolume,
  musicVolume: AUDIO_DEFAULTS.musicVolume,
  mixWithOthers: AUDIO_DEFAULTS.mixWithOthers,
});

// Computed properties helper
export const getAudioComputedProperties = (preferences) => {
  const canPlaySFX = !preferences.globalMute && preferences.soundEffectsEnabled;
  const canPlayMusic = !preferences.globalMute && preferences.musicEnabled;

  return {
    canPlaySFX,
    canPlayMusic,
    effectiveSFXVolume: canPlaySFX ? preferences.soundEffectsVolume : 0,
    effectiveMusicVolume: canPlayMusic ? preferences.musicVolume : 0,
    sfxVolumePercent: Math.round(preferences.soundEffectsVolume * 100),
    musicVolumePercent: Math.round(preferences.musicVolume * 100),
  };
};

// Clamp volume to valid range
export const clampVolume = (volume) => Math.max(0, Math.min(1, volume));

// Load audio preferences from AsyncStorage
export const loadAudioPreferences = async () => {
  try {
    const keys = Object.values(AUDIO_STORAGE_KEYS);
    const pairs = await AsyncStorage.multiGet(keys);

    const stored = {};
    pairs.forEach(([key, value]) => {
      if (value !== null) {
        const shortKey = Object.keys(AUDIO_STORAGE_KEYS).find(
          k => AUDIO_STORAGE_KEYS[k] === key
        );
        if (shortKey) {
          // Parse based on expected type
          if (shortKey === 'sfxVolume' || shortKey === 'musicVolume') {
            stored[shortKey] = parseFloat(value);
          } else {
            stored[shortKey] = JSON.parse(value);
          }
        }
      }
    });

    return {
      globalMute: stored.globalMute ?? AUDIO_DEFAULTS.globalMute,
      soundEffectsEnabled: stored.sfxEnabled ?? AUDIO_DEFAULTS.soundEffectsEnabled,
      musicEnabled: stored.musicEnabled ?? AUDIO_DEFAULTS.musicEnabled,
      soundEffectsVolume: clampVolume(stored.sfxVolume ?? AUDIO_DEFAULTS.soundEffectsVolume),
      musicVolume: clampVolume(stored.musicVolume ?? AUDIO_DEFAULTS.musicVolume),
      mixWithOthers: stored.mixWithOthers ?? AUDIO_DEFAULTS.mixWithOthers,
    };
  } catch (error) {
    Logger.error('AudioPreferences: Failed to load preferences:', error);
    return createDefaultAudioPreferences();
  }
};

// Save audio preferences to AsyncStorage
export const saveAudioPreferences = async (preferences) => {
  try {
    const pairs = [
      [AUDIO_STORAGE_KEYS.globalMute, JSON.stringify(preferences.globalMute)],
      [AUDIO_STORAGE_KEYS.sfxEnabled, JSON.stringify(preferences.soundEffectsEnabled)],
      [AUDIO_STORAGE_KEYS.musicEnabled, JSON.stringify(preferences.musicEnabled)],
      [AUDIO_STORAGE_KEYS.sfxVolume, String(preferences.soundEffectsVolume)],
      [AUDIO_STORAGE_KEYS.musicVolume, String(preferences.musicVolume)],
      [AUDIO_STORAGE_KEYS.mixWithOthers, JSON.stringify(preferences.mixWithOthers)],
    ];

    await AsyncStorage.multiSet(pairs);
  } catch (error) {
    Logger.error('AudioPreferences: Failed to save preferences:', error);
  }
};

// Update a single preference and save
export const updateAudioPreference = async (key, value) => {
  const preferences = await loadAudioPreferences();
  const updated = { ...preferences, [key]: value };
  await saveAudioPreferences(updated);
  return updated;
};
